### DISCUSSION TITLE
Controlling anti-aliasing when rendering to surface that is not the same size as the window

### DISCUSSION THREAD

**Discussion by casey** [gfx-rs/wgpu#6636] [Q&A] [answered]
I'd like to be able to render to a lower resolution and display it in a pixelated fashion, with nearest neighbor interpolation.

I have a resize function that configures the surface and sets up intermediate rendering targets. When I ignore the window size and use 32x32 pixels, it winds up being blurry. This is on MacOS. See below `<details>` elements for the resize function and output in both cases.

<details>
  <summary>actual size</summary>

```rust
  pub(crate) fn resize(&mut self, size: PhysicalSize<u32>) {
    self.config.width = size.width.max(1);
    self.config.height = size.height.max(1);
    self.surface.configure(&self.device, &self.config);
    for target in self.targets.iter_mut() {
      *target = Target::new(
        &self.bind_group_layout,
        &self.config,
        &self.device,
        &self.sampler,
        self.texture_format,
        &self.uniform_buffer,
      );
    }
  }
```

![Screenshot 2024-12-01 at 5 09 26 PM](https://github.com/user-attachments/assets/5198c69b-f143-4d3b-96c1-5c6c71b68b0c)

</details>

<details>
  <summary>32x32</summary>

```rust
  pub(crate) fn resize(&mut self, size: PhysicalSize<u32>) {
    self.config.width = 32;
    self.config.height = 32;
    self.surface.configure(&self.device, &self.config);
    for target in self.targets.iter_mut() {
      *target = Target::new(
        &self.bind_group_layout,
        &self.config,
        &self.device,
        &self.sampler,
        self.texture_format,
        &self.uniform_buffer,
      );
    }
  }
```

![Screenshot 2024-12-01 at 5 09 42 PM](https://github.com/user-attachments/assets/a0357c11-eb2e-4d60-ace6-b4ee99bdb9a8)

</summary>

I looked at the `wgpu` docs, but couldn't find anything that looked like it might control whether or not the surface is displayed in an anti-aliased fashion. I also looked at the WebGPU spec, but couldn't find anything there either.

Is this something that can be controlled from `wgpu`/WebGPU? Or is this something which is up to the underlying surface/windowing implementation? Perhaps it can be controlled by being more explicit with `winit` initializing?

**Comment by cwfitzgerald** [comment, marked as answer]
The expected flow is to render to a literal 32x32 image, then render that image yourself to the swapchain of normal size.

Right now it's technically not defined the behavior if your swapchain is the wrong size.

**Reply by casey** [reply]
Thanks for the answer!
